feat: add `FileUpload` component

snitin315 opened the pull request:
## Description

- API - #2010 
- Tests - #2040 

<!-- Briefly explain the purpose of your PR -->

<!-- List the specific changes made, consider adding screenshots if relevant -->


<!-- Include any relevant details, links to issues, or additional messages -->

saurabhdaware commented on packages/blade/src/components/FileUpload/FileUpload.web.tsx:
<img width="529" alt="image" src="https://github.com/razorpay/blade/assets/30949385/8c5dae54-2eb9-4769-abce-6d7d5834bd69">

The delete icon is going outside when file name is large

snitin315 replied:
It should have been truncated. Can you share the file with me on slack?

<img width="440" alt="Screenshot 2024-02-23 at 2 32 33 PM" src="https://github.com/razorpay/blade/assets/46647141/d3211160-d2fc-447e-a6c8-6b8811532c20">

saurabhdaware replied:
sent file on slack. Maybe because the filename didn't have spacing in my case

saurabhdaware commented on packages/blade/src/components/FileUpload/FileUpload.web.tsx:
Is validation of files needs to be handled on consumer end?

I dragged and dropped a PDF file and it was able to capture and upload it

<img width="529" alt="image" src="https://github.com/razorpay/blade/assets/30949385/622c6797-fe46-4bd2-935a-b72413db5d9e">

snitin315 replied:
I think accept prop is not working for drag & drop. Are you able to select a pdf file from on clicking upload?

saurabhdaware replied:
no it works fine when I click and select. But not working on drag and drop

snitin315 replied:
@saurabhdaware Fixed

anuraghazra commented:
There's a quite a lot of layout shift which is happening after selecting a file, this is happening because after the "Upload GST" label the margin is increasing, ideally margin should stay same. 

https://github.com/razorpay/blade/assets/35374649/eb71436a-8f61-43b1-b83a-6e8338c0f864


On design the margin stays consistent: 

<img width="740" alt="image" src="https://github.com/razorpay/blade/assets/35374649/ff27ad93-6cdf-4c94-933d-e2fb8c9c67ab">

anuraghazra commented:
File name and the Tick icon doesn't look aligned vertically. 

<img height="100" alt="image" src="https://github.com/razorpay/blade/assets/35374649/8b43b719-8b8d-49e8-80ef-76cf376e1e31">

On design: 

<img height="100" alt="image" src="https://github.com/razorpay/blade/assets/35374649/80c1a36c-c69a-4b98-a109-cb63f47ac89b">

anuraghazra commented:
There's a gap in the bottom after the progressbar: 

<img width="350" alt="image" src="https://github.com/razorpay/blade/assets/35374649/f8cfad25-c1ae-40ce-b5eb-216a28685af9">

Gap not present in design: 

<img width="267" alt="image" src="https://github.com/razorpay/blade/assets/35374649/658f8265-64a5-41aa-a615-6c4b42620877">

anuraghazra commented:
In the first storybook example, can you wire up the controls? I cannot play around with different variants like isDisabled.

anuraghazra commented:
Check once with RK on how it looks in mobile: 

<img width="200" alt="image" src="https://github.com/razorpay/blade/assets/35374649/1bdf87b3-1497-44a1-b371-215011b9803c">

Looks a bit too cramped, we may need to add vertical padding in that Upload CTA

anuraghazra commented:
Tried on my android, truncation fails and text overflows for some reason: 

![image](https://github.com/razorpay/blade/assets/35374649/7499953b-dbfa-45f6-a0d5-f5351fa1c854)

anuraghazra commented on packages/blade/src/components/FileUpload/FileUpload.web.tsx:
```diff
@@ -0,0 +1,323 @@
+    }
+
+    if (maxCount && allFiles.length > maxCount) {
+      setErrorMessage(`You can't upload more than ${maxCount} files.`);
```
These messages are hardcoded? 

What if product wants to change these?

snitin315 replied:
These are the default validations messages, giving users control to customize this might complex the logic. We can improve if any use-case comes up. Additional validations & messages can be set via onChange & errorText props.

saurabhdaware replied:
When users pass errorText, will this error message be overriden? or both error messages exist at different places?

snitin315 replied:
`errorText` will override these.

anuraghazra commented on packages/blade/src/components/FileUpload/FileUpload.web.tsx:
```diff
@@ -0,0 +1,323 @@
+            inputProps={{}}
+            style={{
+              cursor: isDisabled ? 'not-allowed' : 'pointer',
+              ...(!isMultiple && selectedFiles.length === 1 ? screenReaderStyles : {}),
```
`!isMultiple && selectedFiles.length === 1`

Factor out and name this conditional with a variable.

snitin315 replied:
done

anuraghazra commented on packages/blade/src/components/FileUpload/FileUpload.web.tsx:
```diff
@@ -0,0 +1,323 @@
+            key={file.id}
+            file={file}
+            onRemove={() => {
+              const newFiles = selectedFiles.filter(({ id }) => id !== file.id);
```
Move this before line 296?

anuraghazra commented on packages/blade/src/components/FileUpload/FileUploadItemIcon.tsx:
```diff
@@ -0,0 +1,101 @@
+    );
+  }
+
+  const FILE_EXTENSION_REGEX = /(?:\.([^.]+))?$/i;
+  const [fileExtension, fileType] = FILE_EXTENSION_REGEX.exec(fileName) ?? [];
```
Do we need regex? 

Can we not split the file in "." and get the last element? 🤔

anuraghazra commented on packages/blade/src/components/Form/Selector/SelectorInput.web.tsx:
```diff
@@ -30,7 +30,8 @@ const getHoverStyles = ({
   return {
     borderColor: borderColor ? getIn(theme, borderColor) : undefined,
-    backgroundColor: getIn(theme, backgroundColor),
+    backgroundColor:
+      backgroundColor === 'transparent' ? 'transparent' : getIn(theme, backgroundColor),
```
We don't need this. 

We have `color.transparent` token in our theme file now.

snitin315 replied:
fixed

anuraghazra commented on packages/blade/src/components/FileUpload/FileUpload.web.tsx:
```diff
@@ -0,0 +1,323 @@
+  useMemo(() => {
+    for (const file of selectedFiles) {
+      if (!file.id) {
+        file.id = `${new Date().getTime().toString()}${Math.floor(Math.random() * 1000000)}`;
```
We are mutating the files without setState?

snitin315 replied:
If we use setState it will go in an infinite re-render loop.

snitin315 replied:
No, hooks can't be called inside a callback.

anuraghazra commented on packages/blade/src/components/FileUpload/FileUpload.web.tsx:
```diff
@@ -0,0 +1,323 @@
+  const showHelpText = !showError && helpText;
+  const accessibilityText =
+    accessibilityLabel ?? `,${showError ? errorText : ''} ${showHelpText ? helpText : ''}`;
+  const { inputId, labelId, helpTextId, errorTextId } = useFormId('fileinput');
```
```suggestion
  const { inputId, labelId, helpTextId, errorTextId } = useFormId('fileuploadinput');
```

anuraghazra commented on packages/blade/src/components/FileUpload/FileUpload.web.tsx:
```diff
@@ -0,0 +1,323 @@
+import { getHintType } from '~components/Input/BaseInput/BaseInput';
+import { makeAccessible } from '~utils/makeAccessible';
+
+const MemoizedFileUploadItem = memo(FileUploadItem);
```
Why do we need to memoize it here? 

Can we not just export a memoized FileInput if this is needed?

snitin315 replied:
yes, done.

snitin315 commented:
> File name and the Tick icon doesn't look aligned vertically.

Fixed

> In the first storybook example, can you wire up the controls? I cannot play around with different variants like isDisabled.


Done

> Check once with RK on how it looks in mobile:

Updated 

>Tried on my android, truncation fails and text overflows for some reason:

Can you try again? I think this is the same as https://github.com/razorpay/blade/pull/2016#discussion_r1500426797

>There's a gap in the bottom after the progressbar:

Fixed.

@anuraghazra

snitin315 commented on packages/blade/src/components/FileUpload/StyledFileUploadItemWrapper.tsx:
```diff
@@ -0,0 +1,36 @@
+      display: 'flex',
+      justifyContent: 'space-between',
+      borderStyle: 'solid',
+      minHeight: '56px',
```
fixed

anuraghazra commented:
> File name and the Tick icon doesn't look aligned vertically.

Weird, cannot find any stories which shows the tick mark anymore, Unable to confirm this. 

> There's a gap in the bottom after the progressbar:

The file upload progress [story](https://61c19ee8d3d282003ac1d81c-ekmmhfqlla.chromatic.com/?path=/story/components-fileupload--auto-file-upload-with-progress) doesn't show the progressbar anymore upon selecting the file, Unable to confirm this.

anuraghazra commented:
this [one](https://github.com/razorpay/blade/pull/2016#issuecomment-1960989807) still exists.

anuraghazra commented:
labelPosition: left looks broken: 

<img width="514" alt="image" src="https://github.com/razorpay/blade/assets/35374649/56e526a5-79fd-477b-9d00-c4b8a4e2f85e">

snitin315 commented:
> The file upload progress [story](https://61c19ee8d3d282003ac1d81c-ekmmhfqlla.chromatic.com/?path=/story/components-fileupload--auto-file-upload-with-progress) doesn't show the progressbar anymore upon selecting the file, Unable to confirm this.

My bad! I've fixed the stories now.

>labelPosition: left looks broken:

Currently, It is the same as other input labels.

<img width="502" alt="Screenshot 2024-02-26 at 3 28 15 PM" src="https://github.com/razorpay/blade/assets/46647141/89c766d7-c0c6-4c8b-bee9-11e005fa2f82">

// @anuraghazra

saurabhdaware commented on packages/blade/src/components/FileUpload/FileUpload.web.tsx:
```diff
@@ -0,0 +1,367 @@
+  const isOneFileSelectedWithSingleUpload = !isMultiple && selectedFiles.length === 1;
+  const inputLabelPosition = platform === 'onMobile' ? 'top' : labelPosition;
+  const isLabelLeftPositioned = inputLabelPosition === 'left';
+  const willRenderHintText = isOneFileSelectedWithSingleUpload
+    ? false
+    : Boolean(helpText) || Boolean(errorMessage);
```
We dont render helpText in mutli upload?

snitin315 replied:
We do, fixed.

saurabhdaware commented on packages/blade/src/components/FileUpload/FileUpload.web.tsx:
```diff
@@ -0,0 +1,367 @@
+          )}
+        </BaseBox>
+      </BaseBox>
+      {/* the magic number 136 is basically max-width of label i.e 120 and then right margin i.e 16 which is the spacing between label and input field */}
```
Move it constant maybe?

snitin315 replied:
It's used in Baseinpu & OTPInput as well, let's refactor this separately.

saurabhdaware commented on packages/blade/src/components/FileUpload/FileUpload.web.tsx:
```diff
@@ -0,0 +1,367 @@
+  };
+
+  const handleInputChange = (event: React.ChangeEvent<HTMLInputElement>): void => {
+    const inputFiles = Array.from(event.target.files ?? []);
```
is it not already an array?

snitin315 replied:
No, [`FileList`](https://developer.mozilla.org/en-US/docs/Web/API/FileList) is not an array.

saurabhdaware commented on packages/blade/src/components/FileUpload/FileUpload.web.tsx:
```diff
@@ -0,0 +1,367 @@
+                  display="flex"
+                  justifyContent="center"
+                  alignItems="center"
+                  flexDirection={{ xs: 'column', s: 'row' }}
```
```suggestion
                  flexDirection={{ base: 'column', s: 'row' }}
```

Is this expected btw? so we want it to be flex on smaller phones than normal `s` size phones?

snitin315 replied:
yes,

saurabhdaware commented on packages/blade/src/components/FileUpload/FileUpload.web.tsx:
```diff
@@ -0,0 +1,367 @@
+                    display="flex"
+                    justifyContent="center"
+                    alignItems="center"
+                    flexDirection={{ xs: 'column', s: 'row' }}
```
```suggestion
                    flexDirection={{ base: 'column', s: 'row' }}
```

Better to use `base` wherever you're using responsive props. It acts as a default value when value is not defined

snitin315 replied:
updated.

saurabhdaware commented on packages/blade/src/components/FileUpload/FileUpload.web.tsx:
```diff
@@ -0,0 +1,367 @@
+};
+
+const FileUpload = assignWithoutSideEffects(forwardRef(_FileUpload), {
+  displayName: MetaConstants.FileUpload,
```
It shouldn't be MetaConstants no? we usually have TitleCased names here for displayName and componentId

snitin315 replied:
Correct, fixed.

saurabhdaware commented on packages/blade/src/components/FileUpload/FileUpload.web.tsx:
```diff
@@ -0,0 +1,367 @@
+    }
+
+    if (maxSize && inputFiles.some((file) => file.size > maxSize)) {
+      setErrorMessage('File size exceeded.');
```
do we not need to tell file size exceeded for which file?

snitin315 replied:
Thinking more on this, we don't need this info. There can be multiple files with long names too. We can add it later if requested by users.

saurabhdaware commented on packages/blade/src/components/FileUpload/FileUpload.web.tsx:
```diff
@@ -0,0 +1,367 @@
+    }
+
+    if (uploadType === 'single' && inputFiles.length > 1) {
+      setErrorMessage('You can upload only one file.');
```
Check these error messages with design also once if they are not coming from design already

saurabhdaware commented on packages/blade/src/components/FileUpload/FileUploadItemIcon.tsx:
```diff
@@ -0,0 +1,100 @@
+    case 'xls':
+    case 'xlsx':
+    case 'csv':
+      iconColors.background = globalColors.chromatic.emerald[50];
```
are we using global colors on design also?

saurabhdaware commented on packages/blade/src/components/FileUpload/FileUpload.web.tsx:
```diff
@@ -0,0 +1,367 @@
+
+  const showError = validationState === 'error' || internalValidationState === 'error';
+  const showHelpText = !showError && helpText;
+  const accessibilityText =
```
Shouldn't the label get priority when accesisibilityLabel is not present?

Like right now when announcing component, it is only announcing help text `".png, .jpg" should be uploaded` but there is no context to users on what type of image. That context will be on label no?

<img width="828" alt="image" src="https://github.com/razorpay/blade/assets/30949385/fd81590b-0f4c-41b6-9fc9-09d656590331">

snitin315 replied:
I'll check this.

snitin315 replied:
Fixed.

snitin315 commented on packages/blade/src/components/FileUpload/FileUpload.web.tsx:
```diff
@@ -0,0 +1,367 @@
+  const { inputId, labelId, helpTextId, errorTextId } = useFormId('fileuploadinput');
+
+  const accessibilityProps = makeAccessible({
+    required: Boolean(isRequired),
+    invalid: Boolean(showError),
+    disabled: Boolean(isDisabled),
```
They can be `undefined` too.

snitin315 commented on packages/blade/src/components/Box/BaseBox/types/propsTypes.ts:
```diff
@@ -127,7 +127,7 @@ type CommonBoxVisualProps = MakeObjectResponsive<
     borderRadius: keyof Border['radius'];
     borderWidth: keyof Border['width'];
-    borderColor: BorderColorString<'surface'>;
+    borderColor: BorderColorString<'surface'> | BorderColorString<'interactive'>;
```
We support drag & drop props on Box too so I think we should support it on both.

snitin315 replied:
For now, I refactored code to avoid using `borderColor` prop. but let's allow `interactive` token for Box in separate PR because we support Drag & Drop.

saurabhdaware replied:
makes sense

snitin315 commented:
@anuraghazra Fixed the stories, you can check again.

> this https://github.com/razorpay/blade/pull/2016#issuecomment-1960989807 still exists.

Fixed

> labelPosition: left looks broken:
> 

It's the current behavior of FormLabel for all our inputs. The `*` goes into the next line.
https://github.com/razorpay/blade/blob/fe64b50e12d66a2eae2580f4b2b109e121067c59/packages/blade/src/components/Form/FormLabel.tsx#L113
<img width="547" alt="Screenshot 2024-02-28 at 8 26 07 AM" src="https://github.com/razorpay/blade/assets/46647141/ea55a7ff-e547-4a32-979e-d9d821541fa1">

saurabhdaware commented on packages/blade/src/components/FileUpload/StyledFileUploadItemWrapper.tsx:
```diff
@@ -24,6 +24,7 @@ const StyledFileUploadItemWrapper = styled(BaseBox)<StyledFileUploadItemWrapperP
       transitionProperty: 'background-color',
       transitionTimingFunction: easing,
       transitionDuration: duration,
+      borderColor: getIn(theme.colors, 'interactive.border.neutral.faded'),
```
```suggestion
      borderColor: theme.colors.interactive.border.neutral.faded,
```

This also works in this case right?

snitin315 replied:
Yes, updated.

saurabhdaware commented on packages/blade/src/components/FileUpload/FileUpload.web.tsx:
```diff
@@ -0,0 +1,363 @@
+  );
+};
+
+const FileUpload = assignWithoutSideEffects(forwardRef(_FileUpload), {
```
Do add jsdoc here with basic usage

snitin315 replied:
Done.

saurabhdaware commented:
The examples in stories and examples on stackblitz seem like same examples. Maybe you can keep basic examples in stories and move complex examples in stackblitz because they'll be able to see full code on stackblitz

snitin315 commented:
> The examples in stories and examples on stackblitz seem like same examples. Maybe you can keep basic examples in stories and move complex examples in stackblitz because they'll be able to see full code on stackblitz

Stackblitz takes significant boot-up time hence I've kept both. Users can play around with stories and look at examples if they want to see the implementation.

saurabhdaware commented on packages/blade/src/components/FileUpload/FileUpload.web.tsx:
```diff
@@ -354,6 +354,119 @@ const _FileUpload: React.ForwardRefRenderFunction<BladeElementRef, FileUploadPro
 
+/**
+ * ### FileUpload Component
+ * 
```
This is a little too large for jsdoc 🙈 Its hard to parse through this much code in small dialog box of jsdoc. Just keep `<FileUpload />` component with its props maybe? ignore the state management and surrounding box buttons etc. Doesn't have to be a working example

snitin315 replied:
Updated but in https://github.com/razorpay/blade/pull/2040
